// @uiw/react-amap-tile-layer/cjs/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard")["default"];
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault")["default"];
Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  TileLayerType: true,
  TileLayer: true
};
exports.TileLayerType = exports.TileLayer = void 0;
var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));
var _react = _interopRequireWildcard(require("react"));
var _useTileLayer2 = require("./useTileLayer");
Object.keys(_useTileLayer2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _useTileLayer2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _useTileLayer2[key];
    }
  });
});
var TileLayerType = /*#__PURE__*/function (TileLayerType) {
  TileLayerType["SATELLITE"] = "satellite";
  TileLayerType["TRAFFIC"] = "traffic";
  TileLayerType["ROADNET"] = "roadnet";
  return TileLayerType;
}({});
exports.TileLayerType = TileLayerType;
var TileLayer = /*#__PURE__*/_react["default"].forwardRef(function (props, ref) {
  var _useTileLayer = (0, _useTileLayer2.useTileLayer)(props),
    tileLayer = _useTileLayer.tileLayer;
  (0, _react.useImperativeHandle)(ref, function () {
    return (0, _objectSpread2["default"])((0, _objectSpread2["default"])({}, props), {}, {
      tileLayer: tileLayer
    });
  }, [tileLayer]);
  return null;
});
exports.TileLayer = TileLayer;
